### ISSUE TITLE
fuzz testing monitor hex0 encoding shows CALLI failure on 64 bit register, no optimizations

### ISSUE THREAD

**Issue by markjenkins** [markjenkins/knightpies#2]
188c3fe77730400a939f85e4f93746bec383901a enhances the hex0 fuzzing test suite by adding register size and register size optimization variations. This demonstrates an interesting failure case, Hex0FuzzTest64 triggers knightdecodeutil.OutsideOfWorldException after the invocation of CALLI. The other variations (16 bit, 32bit, with and without optimized versions, and 64 bit optimize) work fine.

CALLI generic (not register size optimized) was introduced in 3df99e211371a075c59ebeb5e552b30ad27bb312 and modified further by f047215068fe6cc1e1c619f9d2e9f1604961dba9 .

The optimized version was introduced in 44952611c62c19ef8fe624c98abb5db9f4fb6dc9 and modified in f047215068fe6cc1e1c619f9d2e9f1604961dba9.

**Comment by markjenkins** [comment]
There's a good chance CALLI isn't the culprit. I rewrote the generic CALLI to import and delegate to knightinstructions64.CALLI when register size is 64 bits. This doesn't allow the test case to pass like it does when registersize is 64 bits and register size optimized version of instructions are enabled. This suggests to me another instruction is responsible for messing things up. A parrallell run  between the python and C knight implementations may be required to catch the instruction that puts things in a bad state.

**Comment by markjenkins** [comment]
Work in the M1py-wip branch has paid off. I used the under development M1.py to dump the list of DEFINED macros used in stage0_monitor.s . Then I looked at which of those has bit optimized versions.

To pass the test test_hex0_fuzz.Hex0FuzzTest64 I needed to use the 64 bit optimized implementations of both CALLI and RET. Reverting either one of them back to the generic versions would reproduce the issue.

This is too much of a kludge to commit to master. I suspect an endianess issue in the bytes being written to and popped off the stack by these instructions. It's possible that both 64 bit versions of CALLI and RET are actually wrong but work correctly together!

Investigating further.
